File: guess_card.go
// guess_card.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"path/filepath"
	"strings"
	"time"
)

const (
	reset  = "\033[0m"
	red    = "\033[91m"
	green  = "\033[92m"
	yellow = "\033[93m"
	blue   = "\033[94m"
	cyan   = "\033[96m"
	magenta= "\033[95m"
	bold   = "\033[1m"
)

func colorize(text, color string) string {
	return color + text + reset
}

var suits = []string{"пики", "червы", "бубны", "трефы"}
var ranks = []string{"2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"}
var suitColors = map[string]string{
	"пики": blue, "червы": red, "бубны": red, "трефы": green,
}

type Record struct {
	Record int `json:"record"`
}

type Game struct {
	mode      string
	record    int
	recordFile string
	target    [2]string // [suit, rank]
	attempts  int
	gameOver  bool
}

func NewGame(mode string) *Game {
	g := &Game{mode: mode}
	g.recordFile = filepath.Join(os.Getenv("HOME"), ".guess_card_record.json")
	g.loadRecord()
	return g
}

func (g *Game) loadRecord() {
	data, err := os.ReadFile(g.recordFile)
	if err != nil {
		g.record = -1
		return
	}
	var rec Record
	err = json.Unmarshal(data, &rec)
	if err != nil {
		g.record = -1
	} else {
		g.record = rec.Record
	}
}

func (g *Game) saveRecord(rec int) {
	data, _ := json.Marshal(Record{Record: rec})
	os.WriteFile(g.recordFile, data, 0644)
}

func (g *Game) chooseCard() {
	rand.Seed(time.Now().UnixNano())
	s := rand.Intn(len(suits))
	r := rand.Intn(len(ranks))
	g.target = [2]string{suits[s], ranks[r]}
}

func (g *Game) getSuit(name string) string {
	for _, s := range suits {
		if strings.EqualFold(s, name) {
			return s
		}
	}
	return ""
}

func (g *Game) getRank(name string) string {
	for _, r := range ranks {
		if strings.EqualFold(r, name) {
			return r
		}
	}
	return ""
}

func (g *Game) displayCard(suit, rank string) string {
	col := suitColors[suit]
	return colorize(rank+" "+suit, col)
}

func (g *Game) play() {
	fmt.Println(colorize("🃏 Добро пожаловать в 'Угадай карту'!", bold))
	fmt.Println("Компьютер загадал карту. Попробуйте угадать!")
	if g.mode == "full" {
		fmt.Println("Вам нужно угадать и масть, и достоинство.")
	} else if g.mode == "suit" {
		fmt.Println("Вам нужно угадать только масть.")
	} else {
		fmt.Println("Вам нужно угадать только достоинство.")
	}
	if g.record != -1 {
		fmt.Printf("Текущий рекорд: %d попыток.\n", g.record)
	}
	fmt.Println("Для выхода введите 'q'.\n")

	g.chooseCard()
	g.attempts = 0
	scanner := bufio.NewScanner(os.Stdin)

	for !g.gameOver {
		g.attempts++
		if g.mode == "full" {
			fmt.Print("Введите масть (пики/червы/бубны/трефы): ")
			scanner.Scan()
			suitInput := strings.TrimSpace(scanner.Text())
			if strings.ToLower(suitInput) == "q" {
				fmt.Println("Выход.")
				return
			}
			fmt.Print("Введите достоинство (2-10, J, Q, K, A): ")
			scanner.Scan()
			rankInput := strings.TrimSpace(scanner.Text())
			if strings.ToLower(rankInput) == "q" {
				fmt.Println("Выход.")
				return
			}
			suit := g.getSuit(suitInput)
			rank := g.getRank(rankInput)
			if suit == "" || rank == "" {
				fmt.Println(colorize("Неверный ввод. Попробуйте снова.", red))
				g.attempts--
				continue
			}
			if suit == g.target[0] && rank == g.target[1] {
				fmt.Println(colorize("🎉 Поздравляем! Вы угадали карту!", green))
				fmt.Printf("Загадана была: %s\n", g.displayCard(g.target[0], g.target[1]))
				g.gameOver = true
			} else {
				if suit == g.target[0] {
					fmt.Println(colorize("Масть угадана верно!", green))
				} else {
					fmt.Println(colorize("Масть не угадана.", red))
				}
				if rank == g.target[1] {
					fmt.Println(colorize("Достоинство угадано верно!", green))
				} else {
					fmt.Println(colorize("Достоинство не угадано.", red))
				}
			}
		} else if g.mode == "suit" {
			fmt.Print("Введите масть (пики/червы/бубны/трефы): ")
			scanner.Scan()
			suitInput := strings.TrimSpace(scanner.Text())
			if strings.ToLower(suitInput) == "q" {
				fmt.Println("Выход.")
				return
			}
			suit := g.getSuit(suitInput)
			if suit == "" {
				fmt.Println(colorize("Неверная масть.", red))
				g.attempts--
				continue
			}
			if suit == g.target[0] {
				fmt.Println(colorize("🎉 Поздравляем! Вы угадали масть!", green))
				fmt.Printf("Загадана была: %s\n", g.displayCard(g.target[0], g.target[1]))
				g.gameOver = true
			} else {
				fmt.Println(colorize("Неверно. Попробуйте снова.", red))
			}
		} else { // rank
			fmt.Print("Введите достоинство (2-10, J, Q, K, A): ")
			scanner.Scan()
			rankInput := strings.TrimSpace(scanner.Text())
			if strings.ToLower(rankInput) == "q" {
				fmt.Println("Выход.")
				return
			}
			rank := g.getRank(rankInput)
			if rank == "" {
				fmt.Println(colorize("Неверное достоинство.", red))
				g.attempts--
				continue
			}
			if rank == g.target[1] {
				fmt.Println(colorize("🎉 Поздравляем! Вы угадали достоинство!", green))
				fmt.Printf("Загадана была: %s\n", g.displayCard(g.target[0], g.target[1]))
				g.gameOver = true
			} else {
				fmt.Println(colorize("Неверно. Попробуйте снова.", red))
			}
		}
	}
	fmt.Printf("Количество попыток: %d\n", g.attempts)
	if g.record == -1 || g.attempts < g.record {
		g.saveRecord(g.attempts)
		fmt.Println(colorize(fmt.Sprintf("🏆 Новый рекорд: %d попыток!", g.attempts), yellow))
	}
}

func main() {
	mode := "full"
	showRecord := false
	for _, arg := range os.Args[1:] {
		switch arg {
		case "full", "suit", "rank":
			mode = arg
		case "-r", "--record":
			showRecord = true
		case "-h", "--help":
			fmt.Println("Usage: guess_card [full|suit|rank] [-r]")
			return
		}
	}
	if showRecord {
		recFile := filepath.Join(os.Getenv("HOME"), ".guess_card_record.json")
		data, err := os.ReadFile(recFile)
		if err != nil {
			fmt.Println("Рекордов пока нет.")
			return
		}
		var rec Record
		if err := json.Unmarshal(data, &rec); err != nil {
			fmt.Println("Рекордов пока нет.")
			return
		}
		fmt.Printf("Рекорд: %d попыток\n", rec.Record)
		return
	}
	game := NewGame(mode)
	game.play()
}

File: guess_card.rb
#!/usr/bin/env ruby
# guess_card.rb
# encoding: UTF-8

require 'json'
require 'fileutils'

COLORS = {
  reset: "\e[0m",
  red: "\e[91m",
  green: "\e[92m",
  yellow: "\e[93m",
  blue: "\e[94m",
  cyan: "\e[96m",
  magenta: "\e[95m",
  bold: "\e[1m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

SUITS = ['пики', 'червы', 'бубны', 'трефы']
RANKS = ['2','3','4','5','6','7','8','9','10','J','Q','K','A']
SUIT_COLORS = {'пики'=>'blue', 'червы'=>'red', 'бубны'=>'red', 'трефы'=>'green'}

class GuessCardGame
  attr_reader :mode, :record_file, :record, :deck, :target, :attempts, :game_over

  def initialize(mode='full')
    @mode = mode
    @record_file = File.join(Dir.home, '.guess_card_record.json')
    @record = load_record
    @deck = SUITS.product(RANKS)
    @target = nil
    @attempts = 0
    @game_over = false
  end

  def load_record
    return nil unless File.exist?(@record_file)
    data = JSON.parse(File.read(@record_file))
    data['record'] rescue nil
  end

  def save_record(attempts)
    File.write(@record_file, JSON.pretty_generate({ 'record' => attempts }))
  end

  def choose_card
    @target = @deck.sample
  end

  def get_suit(name)
    SUITS.find { |s| s.casecmp(name).zero? }
  end

  def get_rank(name)
    RANKS.find { |r| r.casecmp(name).zero? }
  end

  def display_card(suit, rank)
    col = SUIT_COLORS[suit] || 'reset'
    colorize("#{rank} #{suit}", col)
  end

  def play
    puts colorize("🃏 Добро пожаловать в 'Угадай карту'!", :bold)
    puts "Компьютер загадал карту. Попробуйте угадать!"
    if @mode == 'full'
      puts "Вам нужно угадать и масть, и достоинство."
    elsif @mode == 'suit'
      puts "Вам нужно угадать только масть."
    else
      puts "Вам нужно угадать только достоинство."
    end
    puts "Текущий рекорд: #{@record} попыток." if @record
    puts "Для выхода введите 'q'.\n"

    choose_card
    @attempts = 0

    until @game_over
      @attempts += 1
      if @mode == 'full'
        print "Введите масть (пики/червы/бубны/трефы): "
        suit_input = gets.chomp.strip
        break if suit_input.downcase == 'q'
        print "Введите достоинство (2-10, J, Q, K, A): "
        rank_input = gets.chomp.strip
        break if rank_input.downcase == 'q'
        suit = get_suit(suit_input)
        rank = get_rank(rank_input)
        unless suit && rank
          puts colorize("Неверный ввод. Попробуйте снова.", :red)
          @attempts -= 1
          next
        end
        if suit == @target[0] && rank == @target[1]
          puts colorize("🎉 Поздравляем! Вы угадали карту!", :green)
          puts "Загадана была: #{display_card(@target[0], @target[1])}"
          @game_over = true
        else
          puts suit == @target[0] ? colorize("Масть угадана верно!", :green) : colorize("Масть не угадана.", :red)
          puts rank == @target[1] ? colorize("Достоинство угадано верно!", :green) : colorize("Достоинство не угадано.", :red)
        end
      elsif @mode == 'suit'
        print "Введите масть (пики/червы/бубны/трефы): "
        suit_input = gets.chomp.strip
        break if suit_input.downcase == 'q'
        suit = get_suit(suit_input)
        unless suit
          puts colorize("Неверная масть.", :red)
          @attempts -= 1
          next
        end
        if suit == @target[0]
          puts colorize("🎉 Поздравляем! Вы угадали масть!", :green)
          puts "Загадана была: #{display_card(@target[0], @target[1])}"
          @game_over = true
        else
          puts colorize("Неверно. Попробуйте снова.", :red)
        end
      else # rank
        print "Введите достоинство (2-10, J, Q, K, A): "
        rank_input = gets.chomp.strip
        break if rank_input.downcase == 'q'
        rank = get_rank(rank_input)
        unless rank
          puts colorize("Неверное достоинство.", :red)
          @attempts -= 1
          next
        end
        if rank == @target[1]
          puts colorize("🎉 Поздравляем! Вы угадали достоинство!", :green)
          puts "Загадана была: #{display_card(@target[0], @target[1])}"
          @game_over = true
        else
          puts colorize("Неверно. Попробуйте снова.", :red)
        end
      end
    end
    puts "Количество попыток: #{@attempts}"
    if @record.nil? || @attempts < @record
      save_record(@attempts)
      puts colorize("🏆 Новый рекорд: #{@attempts} попыток!", :yellow)
    end
  end
end

def main
  mode = 'full'
  show_record = false
  ARGV.each do |arg|
    case arg
    when 'full', 'suit', 'rank' then mode = arg
    when '-r', '--record' then show_record = true
    when '-h', '--help'
      puts "Usage: ruby guess_card.rb [full|suit|rank] [-r]"
      return
    end
  end
  if show_record
    rec_file = File.join(Dir.home, '.guess_card_record.json')
    if File.exist?(rec_file)
      data = JSON.parse(File.read(rec_file))
      puts "Рекорд: #{data['record'] || 'нет'} попыток"
    else
      puts "Рекордов пока нет."
    end
    return
  end
  game = GuessCardGame.new(mode)
  game.play
end

main if __FILE__ == $0
